### ISSUE TITLE
Incorrect git status

### ISSUE THREAD

**Issue by xiaoxipang** [romkatv/powerlevel10k#1428]
Thanks for this gorgeous theme and easy-to-use configure command!

I pull a repo from my organization and see several errors when I use them:
1. After I cloned the repo, I immediately check the status with `git status`, the message show
    ```
    On branch master
    Your branch is up to date with 'origin/master'.
    
    nothing to commit, working tree clean
    ```
    However, the git status show this:
    ![image](https://user-images.githubusercontent.com/15240273/120258484-221e6b80-c260-11eb-84e3-4463cbc2c281.png)

2. Then when I try to create a new branch, the git status doesn't show at all
![image](https://user-images.githubusercontent.com/15240273/120258934-f18b0180-c260-11eb-86f0-f2017d17caff.png)

On the other hand, I tried to create a new folder in my local and everything works fine
![image](https://user-images.githubusercontent.com/15240273/120259270-a58c8c80-c261-11eb-9b06-8bacd58bf696.png)

It looks like something is off in that special repo, though I have no clue what params will impact the git status display?
Any suggestions for the investigation point, like a specific file in .git folder I should look into that my impact the `VCS_STATUS_NUM_STAGED` or even the whole git status prompt section?

System: MacOS Big Sur Version 11.4

**Comment by romkatv** [comment]
Please do the following:

1. Run `GITSTATUS_LOG_LEVEL=DEBUG zsh`
2. `cd` into the git repository that's missing git status in prompt.
3. Run `less $GITSTATUS_DAEMON_LOG_POWERLEVEL9K` and look at the end for errors.

If you see "unsupported extension name" in there, look at the name of the extension. If it's related to sparse checkouts, see https://github.com/romkatv/powerlevel10k/issues/816#issuecomment-643475021. If it's not related to sparse checkouts, please let me know the name of the extension. Either way, if the extension is not crucial for your workflow, try disabling it.

**Comment by xiaoxipang** [comment]
@romkatv Thanks for the instruction
I'm using the oh-my-zsh, so I'm not sure if I did it correctly. I added the `GITSTATUS_LOG_LEVEL=DEBUG` in my .zshrc
Then I `cd` into the repo and checked the log and found no error relating to extension. Though there are some error:
![image](https://user-images.githubusercontent.com/15240273/120356932-dce75180-c2d2-11eb-9341-15f0b0b0679b.png)
Then I switched to master branch and got these logs
![image](https://user-images.githubusercontent.com/15240273/120357378-60a13e00-c2d3-11eb-8c6c-732be74953e3.png)

Is the .git folder broken? Or a problem with the theme?

Again, when I use `git status` in both branch, it showed `nothing to commit, working tree clean`

**Comment by romkatv** [comment]
Please attach (or send me by email) `/Users/<REDACTED>/.git/packed-refs`.

**Comment by xiaoxipang** [comment]
@romkatv 
Thanks for the quick reply again! I thought I resolved it after you pointed me to the `packed-refs`. I did a quick try on different repos and the above issue only happened on huge repos with more than 200 branches.
I ran the `git pack-refs --all` for the problematic repos and now it shows git status correctly in my local branches.

**Comment by romkatv** [comment]
Damn :-( Now I won't be able to fix the bug. Is it perchance possible you can recover the problematic packed-refs?

**Comment by xiaoxipang** [comment]
Don't worry, it is only a temporary fix. When I re-clone the same repo, it will have the same exact problem. And I found a public repo to reproduce the problem: https://github.com/Shopify/ruby
Let me know if it is working or not.

**Comment by romkatv** [comment]
Here's what I've tried:

```zsh
git clone https://github.com/Shopify/ruby.git
cd ruby
```

Once I ran these commands, I've got "master" in prompt (as expected).

What happens if *you* run these commands? Do you get "master" in prompt? If not, could you send me `packed-refs` that you get, plus the output of `locale`, `git --version` and `git config --global -l`?

**Comment by romkatv** [comment]
Thanks for sending a clone of the public ruby repo from your machine. This allowed me to track down the bug.

The issue was caused by an unusual but valid format of your local repository. Firstly, there is no `.git/refs/heads/master` and the reference is encoded in `.git/packed-refs`. This on its own wouldn't break `gitstatusd` if `.git/packed-refs` were sorted but it isn't. I wonder how you managed to create a repository of this kind.

I've fixed the bug but haven't released a new version of gitstatusd yet. I was delaying the previous release in the hope that I would incorporate this fix in it but gave up waiting yesterday before you sent me your local ruby repo. This will now have to wait for the next release. There is no ETA.

In the meantime you can run `git pack-refs` to create a regular-looking repository. This will allow gitstatusd to do its job and will also make `git` faster.

**Comment by Jmennius** [comment]
Hi there, it looks like I have a similar issue.
This happens in a worktree of the repo (doesn't happen in the repo itself).
Can I try somehow a fixed version?

P.S. Thanks for your awesome work!

**Comment by romkatv** [comment]
> Hi there, it looks like I have a similar issue.
> This happens in a worktree of the repo (doesn't happen in the repo itself).

Doesn't sound like the same issue.

> Can I try somehow a fixed version?

`git clone` the gitstatus repo and build gitstatusd. Instructions are in the readme.

**Comment by Jmennius** [comment]
> > Hi there, it looks like I have a similar issue.
> > This happens in a worktree of the repo (doesn't happen in the repo itself).
> 
> Doesn't sound like the same issue.


Yeah, likely... there is also a submodule and although `git status` doesn't report anything - `git submodule` does (and I only have a single file reported as unstaged).
Sorry for the noise, most likely my issue (although confusing).

**Comment by romkatv** [comment]
gitstatus uses the equivalent of `git status --ignore-submodules=dirty`. Does that this command show anything?

You can look at the gitstatusd log to see which file is reported as changed. See https://github.com/romkatv/powerlevel10k/issues/1428#issuecomment-851904085

**Comment by Jmennius** [comment]
> gitstatus uses the equivalent of `git status --ignore-submodules=dirty`. Does that this command show anything?
> 
> You can look at the gitstatusd log to see which file is reported as changed. See [#1428 (comment)](https://github.com/romkatv/powerlevel10k/issues/1428#issuecomment-851904085)

TLDR It was a git submodule configuration issue, please disregard. 

The `.gitmodules` configuration was broken (I've moved the submodule in the repository and changed remotes a few days ago - something went wrong apparently).
`git submodule` would report status for one path and an error for another `fatal: no submodule mapping found in .gitmodules for path '<path>'`, `git status` was silent (?).
I've tried using broken `.gitmodules` again to check what `git status --ignore-submodules=dirty` would report but couldn't reproduce the issue.

**Comment by romkatv** [comment]
Thanks for the update. I'm glad you've sorted it out.

**Comment by toby-griffiths** [comment]
I'm experiencing this issue on my repo as well.  I have a single submodule that is initialise inside the repo.  Neither the main repo, nor the submodule have any changes, however Powerline10k is reporting 1 unstaged file & 1 untracked file.

I've tried running the `git pack-refs --all` command, but it didn't help.

Running these steps…

1. Run GITSTATUS_LOG_LEVEL=DEBUG zsh
1. cd into the git repository that's missing git status in prompt.
1. Run less $GITSTATUS_DAEMON_LOG_POWERLEVEL9K and look at the end for errors.

… I saw no errors reported, however it looks like it could be an issue with the case of a filename, since the same file is mentioned as both unstaged, and untracked, and it's a file we recently corrected the case (from peopleController.php to PeopleController.php).

**Comment by toby-griffiths** [comment]
Sorry, I should have added… if I clone the repo down into a new directory then the status is reported correctly.

**Comment by romkatv** [comment]
@toby-griffiths Doesn't sound like the same issue. As a workaround you can manually fix the case of the file in your repo.

```zsh
mv peopleController.php tmp && mv tmp PeopleController.php
```

**Comment by romkatv** [comment]
> Also of note, the directory that this local clone was originally setup on was a case-insensitive drive, however I've just rebuilt my laptop with a case-sensitive drive

In this case I highly recommend re-cloning your repo. git remembers whether your fs is case-sensitive or not when you clone.

**Comment by toby-griffiths** [comment]
@romkatv Thank you for the advice.  I appreciate not the same, but thought the info might be relevant. Apologies if you would have preferred this as a separate issue.

And thanks for the tip.  I will re-clone my repos now.

**Comment by romkatv** [comment]
I'm going to close this issue to avoid getting more unrelated issues reported via comments (already got two). As I mentioned earlier, the original issue has been fixed but I haven't released a new version of gitstatusd yet and there is no ETA for the next release (I won't do a release just for this fix).

**Comment by toby-griffiths** [comment]
Sorry, for reference, the file in the normal file structure was the correct case all along.  I just thing Git had it remembered as with its previous file name case.

Interestingly, renaming the file & committing, then renaming back & committing does not resolve the issue, which I thought it might.

**Comment by igbanam** [comment]
If you're still facing this issue for large git repos, add this to your git config

```gitconfig
[index]
  skipHash = false
```

Or simply run

```
git config --global index.skipHash false
```

**Comment by romkatv** [comment]
@igbanam That's not the same issue.
